framework/quota/src/org/apache/cloudstack/quota/dao/QuotaTariffDaoImpl.java

```diff
@@ -0,0 +1,137 @@
+            throw new CloudRuntimeException("Unable to find tariff plan by usage type");
+        } finally {
+            TransactionLegacy.open(opendb).close();
+        }
+        if (result.size() > 0) {
+            if (s_logger.isDebugEnabled()){
+                s_logger.debug("QuotaTariffDaoImpl::findTariffPlanByUsageType: " + effectiveDate + "quota type " + quotaType  + " val=" + result.get(0).getCurrencyValue());
+            }
+            return result.get(0);
+        } else {
+            if (s_logger.isDebugEnabled()){
+                s_logger.info("QuotaTariffDaoImpl::findTariffPlanByUsageType: Missing quota type " + quotaType);
+            }
+            return null;
+        }
+    }
+
+    @Override
+    public List<QuotaTariffVO> listAllTariffPlans(final Date effectiveDate) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        List<QuotaTariffVO> tariffs = new ArrayList<QuotaTariffVO>();
+        try (TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB)) {
+            final Filter filter = new Filter(QuotaTariffVO.class, "effectiveOn", false, 0L, 1L);
+            final SearchCriteria<QuotaTariffVO> sc = listAllIncludedUsageType.create();
+            sc.setParameters("onorbefore", effectiveDate);
+            for (Integer quotaType : QuotaTypes.listQuotaTypes().keySet()) {
+                sc.setParameters("quotatype", quotaType);
+                List<QuotaTariffVO> result = search(sc, filter);
+                if (result.size() > 0) {
+                    tariffs.add(result.get(0));
+                    s_logger.info("listAllTariffPlans onorbefore" + effectiveDate + "quota type " + result.get(0).getDescription() + " , effective Date=" + result.get(0).getEffectiveOn() + " val=" + result.get(0).getCurrencyValue());
+                }
+            }
+        } catch (Exception e) {
+            throw new CloudRuntimeException("Unable to list all tariff plans");
+        } finally {
+            TransactionLegacy.open(opendb).close();
+        }
+        return tariffs;
+    }
+
+    @Override
+    public boolean updateQuotaTariff(QuotaTariffVO plan) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        boolean result = false;
+        try (TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB)) {
+            result = this.update(plan.getId(), plan);
+        } catch (Exception e) {
+            throw new CloudRuntimeException("Unable to update quota tariff");
+        } finally {
+            TransactionLegacy.open(opendb).close();
+        }
+        return result;
+    }
+
+    @Override
+    public QuotaTariffVO addQuotaTariff(QuotaTariffVO plan) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        QuotaTariffVO result = null;
+        try (TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB)) {
+            plan.setId(null);
```

Comment:
That's a additional check to ensure it won't end up updating an existing entry. By setting the ID field to null, this will persist or create a new record/row. @abhinandanprateek can comment more?